### ISSUE TITLE
Feature: Plugins

### ISSUE THREAD

**Issue by Multiply** [arunoda/meteor-up-legacy#130]
I would love for a way to hook into your templates, using 'plugins'.

When you run 'setup' I'd like to be able to add something to, say, your upstart template, or maybe run some special command after the app is deployed, kind of like a hook.

My current example would be to run X instances of forever, each with their own environment variable, specific to the current thread you'd like to use.

Kind of like this:

``` bash
initialPort=$PORT
for core in $(seq 1 $(grep -c '^processor' /proc/cpuinfo))
do
    export METEOR_TASKSET=$core
    export PORT=$(expr $core + $initialPort - 1)
    forever -c userdown --minUptime 2000 --spinSleepTime 1000 app/main.js
done
```

If I could accomplish this using a plugin, or a hook, or something like it, I would be able to spawn a Meteor process for each CPU thread, and I could use my package https://atmospherejs.com/multiply/taskset to set the affinty.

I doubt my above solution to multiple CPUs is something you'd like to implement, as it's pretty much linux specific, and as thus, I figure plugin support would be a good idea?

**Comment by arunoda** [comment]
I think plugin or hooks is a good idea to have.
Yeah! Multi cpu support is planned. I am working on load balancer for
meteor using nginx. It's already ready.

But needs to wrap it and build into mup itself. Then we can simply can have
multi cpu and multi server very easily.

On Thursday, September 11, 2014, Jens Ulrich Hjuler Pedersen <
notifications@github.com> wrote:

> I would love for a way to hook into your templates, using 'plugins'.
> 
> When you run 'setup' I'd like to be able to add something to, say, your
> upstart template, or maybe run some special command after the app is
> deployed, kind of like a hook.
> 
> My current example would be to run X instances of forever, each with their
> own environment variable, specific to the current thread you'd like to use.
> 
> Kind of like this:
> 
> for core in $(seq 1 $(grep -c '^processor' /proc/cpuinfo))do
>     export METEOR_TASKSET=$core
>     export PORT=$(expr $core + $PORT - 1)
>     forever -c userdown --minUptime 2000 --spinSleepTime 1000 app/main.jsdone
> 
> If I could accomplish this using a plugin, or a hook, or something like
> it, I would be able to spawn a Meteor process for each CPU thread, and I
> could use my package https://atmospherejs.com/multiply/taskset to set the
> affinty.
> 
> I doubt my above solution to multiple CPUs is something you'd like to
> implement, as it's pretty much linux specific, and as thus, I figure plugin
> support would be a good idea?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/arunoda/meteor-up/issues/130.

## 

---

Arunoda Susiripala
I curate Meteor Weekly - Check it out!
http://meteorhacks.com/meteor-weekly/?utm_source=email-footer&utm_medium=email&utm_campaign=meteorweekly

**Comment by Multiply** [comment]
I'm highly interested in helping out with the multi-cpu thingy. I made tweaks to the upstart script, so it works with my current solution.

Does each Meteor instance contact the main nginx instances, and add themselves to the queue, or how did you achieve it?

**Comment by arunoda** [comment]
Yes. They do send to the main nginx server. There is a project called
serfdom.io. Which can be used for communication.

On Thursday, September 11, 2014, Jens Ulrich Hjuler Pedersen <
notifications@github.com> wrote:

> I'm highly interested in helping out with the multi-cpu thingy. I made
> tweaks to the upstart script, so it works with my current solution.
> 
> Does each Meteor instance contact the main nginx instances, and add
> themselves to the queue, or how did you achieve it?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/arunoda/meteor-up/issues/130#issuecomment-55253382.

## 

---

Arunoda Susiripala

@arunoda http://twitter.com/arunoda
http://gplus.to/arunodahttps://github.com/arunoda
http://www.linkedin.com/in/arunoda

**Comment by Multiply** [comment]
That looks very promising. I wouldn't mind being an early alpha tester. ;)

**Comment by arunoda** [comment]
Awesome. :)

On Thursday, September 11, 2014, Jens Ulrich Hjuler Pedersen <
notifications@github.com> wrote:

> That looks very promising. I wouldn't mind being an early alpha tester. ;)
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/arunoda/meteor-up/issues/130#issuecomment-55253699.

## 

---

Arunoda Susiripala

@arunoda http://twitter.com/arunoda
http://gplus.to/arunodahttps://github.com/arunoda
http://www.linkedin.com/in/arunoda

**Comment by MaazAli** [comment]
Any updates specific to the multi CPU support @arunoda
